protocols/request-response/tests/error_reporting.rs

```diff
@@ -0,0 +1,727 @@
+            .send_request(&peer1_id, Action::TimeoutOnReadResponse);
+
+        loop {
+            match swarm2.select_next_some().await.try_into_behaviour_event() {
+                Ok(request_response::Event::OutboundFailure {
+                    peer,
+                    request_id,
+                    error,
+                }) => {
+                    assert_eq!(peer, peer1_id);
+                    assert_eq!(request_id, req_id);
+                    assert!(matches!(error, OutboundFailure::Timeout));
+                    break;
+                }
+                Ok(ev) => panic!("Peer2: Unexpected event: {ev:?}"),
+                Err(..) => {}
+            }
+        }
+    };
+
+    spawn(swarm1_task);
+    timeout(Duration::from_millis(200), swarm2_task)
+        .await
+        .expect("timed out on waiting TimeoutOnReadResponse");
+
+    assert!(!panic_check_rx.is_closed(), "swarm1_task panicked");
+}
+
+#[async_std::test]
+async fn report_inbound_failure_on_read_request() {
+    let _ = env_logger::try_init();
+
+    let protocols = iter::once((StreamProtocol::new("/test/1"), ProtocolSupport::Full));
+    let cfg = request_response::Config::default();
+
+    let mut swarm1 = Swarm::new_ephemeral(|_| {
+        request_response::Behaviour::<TestCodec>::new(protocols.clone(), cfg.clone())
+    });
+    let peer1_id = *swarm1.local_peer_id();
+
+    let mut swarm2 =
+        Swarm::new_ephemeral(|_| request_response::Behaviour::<TestCodec>::new(protocols, cfg));
+
+    swarm1.listen().await;
+    swarm2.connect(&mut swarm1).await;
+
+    // On panic `panic_check_rx` will be closed
+    let (panic_check_tx, panic_check_rx) = channel::bounded::<()>(1);
+
+    // Server
+    //
+    // Expects no events because `Event::Request` is produced after `read_request`.
+    let swarm1_task = async move {
+        let _panic_check_tx = panic_check_tx;
+
+        loop {
+            match swarm1.select_next_some().await.try_into_behaviour_event() {
+                Ok(ev) => panic!("Peer1: Unexpected event: {ev:?}"),
+                Err(..) => {}
+            }
+        }
+    };
+
+    // Expects io::ErrorKind::UnexpectedEof
+    let swarm2_task = async move {
+        let req_id = swarm2
+            .behaviour_mut()
+            .send_request(&peer1_id, Action::FailOnReadRequest);
+
+        loop {
+            match swarm2.select_next_some().await.try_into_behaviour_event() {
+                Ok(request_response::Event::OutboundFailure {
+                    peer,
+                    request_id,
+                    error,
+                }) => {
+                    assert_eq!(peer, peer1_id);
+                    assert_eq!(request_id, req_id);
+
+                    let error = match error {
+                        OutboundFailure::Io(e) => e,
+                        e => panic!("Peer2: Unexpected error {e:?}"),
+                    };
+
+                    assert_eq!(error.kind(), io::ErrorKind::UnexpectedEof);
```

Comment:
I think I'd prefer if we have the future return `error.kind()` and then we assert outside.

Reply:
That would allow you to create some helper functions that e.g. drive a swarm until an `OutboundFailure` event is emitted. You can then reuse these functions in all tests and assert for a different `ErrorKind`.

Reply:
done